astropy/utils/misc.py

```diff
@@ -922,12 +922,40 @@ def size(self):
     def isscalar(self):
         return self.shape == ()
 
+    def __len__(self):
+        if self.isscalar:
+            raise TypeError("Scalar {0!r} object has no len()"
+                            .format(self.__class__.__name__))
+        return self.shape[0]
+
+    def __bool__(self):  # Python 3
+        """Any instance should evaluate to True, except when it is empty."""
+        return self.size > 0
+
+    def __nonzero__(self):  # Python 2
+        """Any instance should evaluate to True, except when it is empty."""
+        return self.size > 0
+
     def __getitem__(self, item):
-        if not self.shape:
+        if self.isscalar:
             raise TypeError('scalar {0!r} object is not subscriptable.'.format(
                 self.__class__.__name__))
         return self._apply('__getitem__', item)
 
+    def __iter__(self):
+        if self.isscalar:
+            raise TypeError('scalar {0!r} object is not iterable.'.format(
```

Comment:
not really important but I find it more readable if you start a new line and then use format:

```
raise TypeError('scalar {0!r} object is not iterable.'
                ''.format(self.__class__.__name__))
```